### PR TITLE
UI-1 Added full text search with lunr

### PR DESCRIPTION
**Pull request by arcanite24**
> Added full text search for all the blog posts using lunr.js

Technical Details:
- We generate the index cache at build time using Gatsby own NodeJS pipeline (under /gatsby-node.js)
- We're only indexing the title and an excerpt of max 500 characters
- The cache is saved to /static/idx.json then at build time copied to /public/idx.json
- The cache is accesible by anyone under blog.deimos.app/idx.json

### CODE REVIEW

File: gatsby-node.js [hunk 1]
```diff
@@ -33,31 +40,42 @@ exports.createPages = async ({ graphql, actions }) => {
...
+        context: {
+          slug: post.node.fields.slug,
+          previous,
+          next,
+        },
+      });
     });
   });
+
+  fs.writeFileSync(
```
Comment (ElZombieIsra) [inline gatsby-node.js]: Wrap in a "writeFile" util function?

File: gatsby-node.js [hunk 2]
```diff
@@ -33,31 +40,42 @@ exports.createPages = async ({ graphql, actions }) => {
...
+
+    posts.forEach((post, index) => {
+      const previous =
+        index === posts.length - 1 ? null : posts[index + 1].node;
+      const next = index === 0 ? null : posts[index - 1].node;
 
-  posts.forEach((post, index) => {
-    const previous = index === posts.length - 1 ? null : posts[index + 1].node;
-    const next = index === 0 ? null : posts[index - 1].node;
+      _this.add({
```
Comment (ElZombieIsra) [inline gatsby-node.js:62]: For-of/for-in/for to avoid creating _this wrapper?

File: src/components/layout.js [hunk 3]
```diff
@@ -1,13 +1,46 @@
...
+      }
+    }
+  `);
+
   return (
     <div className="w-100 flex flex-col">
-      <Header></Header>
+      <Header
+        posts={data.posts.edges
+          .map(p => ({
```
Comment (ElZombieIsra) [inline src/components/layout.js]: More descriptive variables names

File: src/components/layout.js [hunk 4]
```diff
@@ -1,13 +1,46 @@
...
+            [p.node.id]: { ...p.node.fields, ...p.node.frontmatter },
+          }))
+          .reduce((acc, curr) => ({ ...acc, ...curr }), {})}
```
Comment (ElZombieIsra) [inline src/components/layout.js:42]: Same

File: src/components/search/GlobalSearch.jsx [hunk 5]
```diff
@@ -1,18 +1,108 @@
...
+const idx = lunr.Index.load(indexCache);
+
+const DEBOUNCE_DELAY = 300;
+const MIN_SEARCH_LENGTH = 2;
+
+const search = query => {
+  return idx.search(query);
+};
+
+const GlobalSearch = ({ placeholder = 'Search something', posts }) => {
```
Comment (ElZombieIsra) [inline src/components/search/GlobalSearch.jsx:23]: Are we going to use localizations? If so, we need to map these strings ASAP so they won't become a pain to map later

File: src/components/search/GlobalSearch.jsx [hunk 6]
```diff
@@ -1,18 +1,108 @@
...
+  const [query, setQuery] = useState('');
+  const [value] = useDebounce(query, DEBOUNCE_DELAY);
+  const [results, setResults] = useState(null);
+  const [selected, setSelected] = useState(null);
+  const [currentIndex, setCurrentIndex] = useState(0);
+
+  const upPressed = useKeyPress(['ArrowUp']);
```
Comment (ElZombieIsra) [inline src/components/search/GlobalSearch.jsx]: Maybe write these names into an enum so they won't be misspelled in the future.

File: src/components/search/GlobalSearch.jsx [hunk 7]
```diff
@@ -1,18 +1,108 @@
...
+  const downPressed = useKeyPress(['ArrowDown']);
+  const enterPressed = useKeyPress(['Enter']);
+
+  useEffect(() => {
+    if (upPressed) {
+      const newIndex = currentIndex > 0 ? currentIndex - 1 : results.length - 1;
```
Comment (ElZombieIsra) [inline src/components/search/GlobalSearch.jsx]: Maybe wrap this logic in a "changeIndex" function to make it cleaner?

File: src/components/search/GlobalSearch.jsx [hunk 8]
```diff
@@ -1,18 +1,108 @@
...
+      setSelected(results[newIndex].ref);
+      setCurrentIndex(newIndex);
+    }
+
+    if (downPressed) {
+      const newIndex = currentIndex < results.length - 1 ? currentIndex + 1 : 0;
```
Comment (ElZombieIsra) [inline src/components/search/GlobalSearch.jsx]: Same as above

File: src/components/search/GlobalSearch.jsx [hunk 9]
```diff
@@ -1,18 +1,108 @@
...
+import indexCache from '../../../static/idx.json';
 import './GlobalSearch.scss';
+import { useEffect } from 'react';
+import classnames from 'classnames';
+import useKeyPress from '../../hooks/useKeyPress';
+import { navigate } from 'gatsby';
+
+const idx = lunr.Index.load(indexCache);
+
+const DEBOUNCE_DELAY = 300;
```
Comment (ElZombieIsra) [inline src/components/search/GlobalSearch.jsx:16]: Isn't it better a 1000ms debounce delay time?
Append metric time name to variable name ex: `DEBOUNCE_TIME_IN_MILLISECONDS`

File: src/components/search/GlobalSearch.jsx [hunk 10]
```diff
@@ -1,18 +1,108 @@
...
+    }
+
+    if (downPressed) {
+      const newIndex = currentIndex < results.length - 1 ? currentIndex + 1 : 0;
+      setSelected(results[newIndex].ref);
+      setCurrentIndex(newIndex);
+    }
+
+    if (enterPressed) {
+      const actualPost = results[currentIndex];
```
Comment (ElZombieIsra) [inline src/components/search/GlobalSearch.jsx:53]: Actual or current post?

File: src/components/search/GlobalSearch.jsx [hunk 11]
```diff
@@ -1,18 +1,108 @@
...
+        navigate(actualPost.slug);
+      }
+    }
+  }, [upPressed, downPressed, enterPressed]);
+
+  useEffect(() => {
+    if (value && value.length >= MIN_SEARCH_LENGTH) {
+      const foundPosts = search(value);
+      setResults([
+        ...foundPosts.map(p => ({
```
Comment (ElZombieIsra) [inline src/components/search/GlobalSearch.jsx]: Var names

File: src/components/search/GlobalSearch.jsx [hunk 12]
```diff
@@ -1,18 +1,108 @@
...
+          ...posts[p.ref],
+          ref: p.ref,
+        })),
+        {
+          slug: 'asdasdasd',
```
Comment (ElZombieIsra) [inline src/components/search/GlobalSearch.jsx]: Hardcoded post? Fetch from somewhere else

File: src/components/search/GlobalSearch.jsx [hunk 13]
```diff
@@ -1,18 +1,108 @@
...
+          title: 'Otra wea',
+          ref: '2',
+        },
+      ]);
+      setSelected(foundPosts[0]?.ref);
```
Comment (ElZombieIsra) [inline src/components/search/GlobalSearch.jsx:69]: Is it safe to use null safety operators in js?

File: src/pages/index.js [hunk 14]
```diff
@@ -18,8 +18,6 @@ const BlogIndex = ({ data, location }) => {
   return (
     <Layout location={location} title={siteTitle}>
       <SEO title="All posts" />
-      {/* <p>{JSON.stringify(posts)}</p>
```
Comment (ElZombieIsra) [inline src/pages/index.js:21]: Remove not used code to avoid trash and improve readability

**Review by ElZombieIsra — CHANGES_REQUESTED** [review]
Overall changes and code styling

File: src/components/layout.js (hunk 3 above)
Comment (arcanite24) [reply src/components/layout.js]: change to .post

File: src/components/search/GlobalSearch.jsx (hunk 5 above)
Comment (arcanite24) [reply src/components/search/GlobalSearch.jsx:23]: No we're not using i18n

File: src/components/search/GlobalSearch.jsx (hunk 9 above)
Comment (arcanite24) [reply src/components/search/GlobalSearch.jsx:16]: 300 for UX seems ok

File: gatsby-node.js (hunk 2 above)
Comment (arcanite24) [reply gatsby-node.js:62]: Tested with normal for and Lunr required forEach :$ keeping this for now

File: gatsby-node.js [hunk 15]
```diff
@@ -1,5 +1,15 @@
 const path = require('path');
+const fs = require('fs');
 const { createFilePath } = require('gatsby-source-filesystem');
+const lunr = require('lunr');
+require('lunr-languages/lunr.stemmer.support')(lunr);
+require('lunr-languages/lunr.es')(lunr);
+
+const writeFile = (path, payload) => {
+  fs.writeFileSync(path, payload);
+};
```
Comment (ElZombieIsra) [inline gatsby-node.js:10]: Abstract to a utils folder. Approving for now.

File: gatsby-node.js (hunk 2 above)
Comment (ElZombieIsra) [reply gatsby-node.js:62]: Lil sus, but ok

File: src/components/search/GlobalSearch.jsx (hunk 9 above)
Comment (ElZombieIsra) [reply src/components/search/GlobalSearch.jsx:16]: Const name didn't change(?),

**Review by ElZombieIsra — APPROVED** [review]
Parece basura para mi

**Pull request merged**